### DISCUSSION TITLE
Keep a cached mongoose connection with serverless functions

### DISCUSSION THREAD

**Discussion by vebits** [vercel/community#424] [Help] [answered]
Hi, how can I keep a cached mongoose connection throughout multiple serverless functions?

Today I need to run this block of code of top of every serveless functions that need DB connection.

```
const mongoose = require("mongoose");

const options = {
  useNewUrlParser: true,
  useCreateIndex: true,
  useUnifiedTopology: true
};

const CONNECTION_STRING = "MONGODB_CONNECTION_STRING";

module.exports = () => {
  mongoose.connect(CONNECTION_STRING, options);

  const connection = mongoose.connection;
  connection.once("open", () => {
    console.log("MongoDB database connection established successfully.");
  });

  mongoose.connection.on("error", err => {
    console.log(err);
  });
};
```

But it is cumbersome to always open a new connection for each serverless function, so does anyone have any idea how to solve this?

I have seen the Vercel example with MongoDb, but I couldn't find any useful information there.

**Comment by styfle** [comment, marked as answer]
The [MongoDB Guide](https://vercel.com/guides/deploying-a-mongodb-powered-api-with-node-and-vercel#step-2:-creating-the-api) has an example that shows how to cache the DB connection between function invocations.

Note the global `cachedDb` variable.


```js
const url = require('url')
const MongoClient = require('mongodb').MongoClient

let cachedDb = null // Create cached connection variable

async function connectToDatabase(uri) {
  if (cachedDb) {
    return cachedDb // Prefer cached connection
  }

  const client = await MongoClient.connect(uri, { useNewUrlParser: true })
  const db = await client.db(url.parse(uri).pathname.substr(1))
  cachedDb = db // Cache the database connection
  return db
}

module.exports = async (req, res) => {
  const db = await connectToDatabase(process.env.MONGODB_URI)
  const collection = await db.collection('users')
  const users = await collection.find({}).toArray()
  res.status(200).json({ users })
}
```

**Reply by vebits** [reply]
Ok, but is this connection only cached for THAT specific serverless function or can you cache a db connection and use it with multiple serverless function?

**Reply by styfle** [reply]
Its only cached for the specific serverless function

**Reply by lemmon** [reply]
Works as expected. However, I do encounter problems when I pass `useUnifiedTopology: true` option. Connections keep dropping left and right and I am getting lambda timeouts after 10s. When I drop `useUnifiedTopology` option everything works just fine. All the dependencies are up-to-date. Do you have any suggestions why is this happening? Thanks.

**Reply by CITIZENDOT** [reply]
Could you give similar example for `mongoose` please?

**Reply by andreasntr** [reply]
For those using serverless offline, this behaviour is normal and it depends on how the plugin manages its cache. This trick works if you start you local server with `serverless offline --allowCache`

It's the same for mongoose, just put you connect function in a separate file:

```javascript
let isConnected;

exports.connectToDatabase = () => {
  if (isConnected) {
    console.log('=> using existing database connection');
    return Promise.resolve();
  }

  console.log('=> using new database connection');
  return mongoose.connect(process.env.DATA_DB_URL, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    bufferCommands: false
  })
    .then(db => {
      isConnected = db.connections[0].readyState;
    });
};
```

**Comment by lemmon** [comment]
I came accross a solution that seems to be working fine 🤞:
https://github.com/vercel/next.js/blob/canary/examples/with-mongodb/util/mongodb.js

Aparently the trick is using the `global` variable:
```js
let cached = global.mongo
```

**Reply by diogocsc** [reply]
An alternative can be https://docs.atlas.mongodb.com/best-practices-connecting-from-vercel/

**Reply by kevle1** [reply]
Another option could be using the MongoDB Atlas Data API - it doesn't use up any connections https://www.mongodb.com/docs/atlas/api/data-api/#when-to-use-the-data-api

**Comment by robots4life** [comment]
Mongoose: https://github.com/vercel/next.js/blob/canary/examples/with-mongodb-mongoose/lib/dbConnect.js
MongoDB: https://github.com/vercel/next.js/blob/canary/examples/with-mongodb/lib/mongodb.js

**Reply by Mihai-github** [reply]
HI, thanks. I've used the Mongoose version and seems to work

Logs:

`CONNECTING TO MONGO`
`CACHED CONNECTION OF MONGO`
`CONNECTED TO MONGO`


Also, does anyone figured out something that may others come across:

`For the /api/* folder how can we create only once the connection with mongo and not on every handler function for a request??`
`I've tried finding an API middleware but had no luck finding something ok, at least from my point of view... maybe someone knows?`

**Reply by emhat098** [reply]
Just update the link:
Mongoose: [https://github.com/vercel/next.js/blob/canary/examples/with-mongodb-mongoose/lib/dbConnect.ts](Mongoose)
MongoDB: [https://github.com/vercel/next.js/blob/canary/examples/with-mongodb/lib/mongodb.ts](MongoDB)
